src/librustc/infer/error_reporting/mod.rs

```diff
@@ -966,3 +712,515 @@ impl<'tcx> ObligationCause<'tcx> {
         }
     }
 }
+
+/// Visitor that collects every `Item` imported into the current `TyCtxt`'s scope.
+struct ImportVisitor<'a, 'gcx: 'a + 'tcx, 'tcx: 'a> {
+    tcx: TyCtxt<'a, 'gcx, 'tcx>,
+    /// All the `Item`s from other scopes visible from this `TyCtxt`, as well as their local name.
+    full_imports: FxHashMap<String, String>,
```

Comment:
Why not an  `FxHashMap<DefId, String>`?

Reply:
This is an artifact from the exploratory way I wrote this code, as while working on this I didn't have access to `DefId`s from other crates. Very sensible suggestion.